src/codegen/translate-expression.lisp

```diff
@@ -984,10 +989,28 @@ dictionaries applied."
         :finally (return inner)))
 
 (defun wrap-in-block (inner)
-  "Wrap INNER in a block named for the current function."
+  "Check whether INNER contains any return statements which need a
+block. If so, wrap INNER in a block named for the current function.
+Otherwise, return INNER unchanged.
+
+There is currently no way for any return node to jump outside its
+function's block, but this traversal handles the most general case by
+checking for any return nodes targeting *CURRENT-FUNCTION-NAME* which
+don't occur in an already existing block with a matching label."
   (declare (type node inner)
            (values node &optional))
-  (make-node-block
-   :type (node-type inner)
-   :name *current-function-name*
-   :body inner))
+  (traverse
+   inner
+   (list
+    (action (:traverse node-block node)
+      (unless (eq *current-function-name* (node-block-name node))
+        (funcall *traverse* (node-block-body node)))
+      (values))
+    (action (:before node-return-from node)
+      (when (eq *current-function-name* (node-return-from-name node))
+        (return-from wrap-in-block
```

Comment:
it's funny that we use early exit in a language that doesn't allow early exit via return in a lambda